File: deeplearning_c++/deeplearning_c++/convnet_maxpool_ex1.c
//# include<stdio.h>
//
//int main()
//{
//	int image[5][5] = { {3,2,4,5,6},
//				  {4,5,6,1,2},
//				  {2,3,1,0,2},
//				  {4,2,1,4,3},
//				  {3,4,7,2,1} };
//
//	int kernel[2][2] = { {1,1},{0,1} };
//
//	int feature[4][4] = { {0,0,0,0},
//						{0,0,0,0},
//						{0,0,0,0},
//						{0,0,0,0} };
//
//	for (int i = 0; i < 4; i++) {
//		for (int j = 0; j < 4; j++) {
//			int sum = 0;
//
//			for (int h = 0; h < 2; h++) {
//				for (int w = 0; w < 2; w++) {
//					sum += image[i + h][j + w] * kernel[h][w];
//				}
//			}
//			feature[i][j] += sum;
//			printf("%d\n", feature[i][j]);
//		}
//	}
//
//	int maxpool[2][2] = { {0,0},
//						  {0,0} };
//
//
//	for (int i = 0; i < 2; i++) {
//		for (int j = 0; j < 2; j++) {
//			int sum = -2147483647;
//
//			for (int h = 0; h < 2; h++) {
//				for (int w = 0; w < 2; w++) {
//					if (sum < feature[2*i + h][2*j + w]) {
//						sum = feature[2*i + h][2*j + w];
//					}
//				}
//			}
//			maxpool[i][j] += sum;
//			printf("%d\n", maxpool[i][j]);
//		}
//	}
//
//	// convolution 5*5 & 2*2
//	// maxpooling
//
//	return 0;
//}

File: deep_python/convnet_maxpool_ex1.py
  
image = [[3,2,4,5,6],
[4,5,6,1,2],
[2,3,1,0,2],
[4,2,1,4,3],
[3,4,7,2,1]]

kernel = [[1,1],[0,1]]

feature = [[0,0,0,0] for i in range(4)]

#[[0,0,0,0],
# [0,0,0,0],
# [0,0,0,0],
# [0,0,0,0]]

for h in range(4): # featuremap 크기 기준
    for w in range(4):
        s=0
        for h1 in range(2): # 필터 크기 기준
            for w1 in range(2):
                s += image[h+h1][w+w1] * kernel[h1][w1]
        feature[h][w] += s

print(feature)


maxpool_feature = [[0,0] for i in range(2)]

for y in range(2):
    for u in range(2):
        max = -2147483648 # int 32의 최대 음수값

        for g in range(2):
            for h in range(2):
                if(max <feature[2*y+g][2*u+h]):
                    max = feature[2*y+g][2*u+h]
        maxpool_feature[y][u] +=max
print(maxpool_feature)